Source/WebCore/rendering/StyledMarkedText.cpp

```diff
@@ -203,4 +299,17 @@ Vector<StyledMarkedText> StyledMarkedText::coalesceAdjacentWithEqualDecorations(
     });
 }
 
+void StyledMarkedText::removeOverlappingForegroundRanges(Vector<StyledMarkedText>& markedTexts)
```

Comment:
I may be misunderstanding but the code in this function doesn't seem to make sense. It seems:
1. You create a HashSet containing all the [startOffset, endOffset] of the markedTexts that don't have a highlightName
2. Then you remove from markedTexts all the texts that have don't have a highlightName and whose [startOffset, endOffset] is in the HashSet.

It seems it would remove all markedTexts that don't have a highlight name no?

Reply:
If 2 ranges overlap, I am not sure if you want to remove all of them or only the second one. My code above assumes you want to remove all that overlaps except for the first one.

If you want to remove all the ranges that overlap, you could do something like:
```
HashCountedSet<std::pair<unsigned, unsigned>> ranges;
for (auto& markedText : markedTexts) {
    if (!markedText.highlightName.isNull())
        ranges.add(std::pair { markedText.startOffset, markedText.endOffset });
}
markedTexts.removeAllMatching([&](auto& markedText) {
    return markedText.highlightName.isNull() && ranges.count(std::pair { markedText.startOffset, markedText.endOffset }) > 1;
});
```

Either way, removing all markedText which doesn't have a highlight name doesn't seem what you intended to do based on your comment inside the function where you talk about overlap.

Also note that your comment mentions "overlap", I think 2 ranges can have overlap without necessarily be identical.
I didn't check the spec but it'd be good to check what the exact behavior is supposed to be. Maybe we want to merge overlapping ranges into a single range by doing a union for e.g.

Reply:
I have a `HashSet` containing all ranges of `MarkedTexts` that _do_ have a highlight name (aka highlight ranges). Then, I remove all the MarkedText that are identical with those specific ranges. This is because the vector contains MarkedTexts that are the original text and MarkedTexts of highlights that contain the word the highlight covers. Thus, it would paint the text of the highlight and then another `MarkedText` with identical range would paint over that same word. This caused unwanted behavior if the custom highlight specified a different text color. At the time, I did not have a sophisticated coalescing styles function and I sorted the vector so all non-highlights would be at the end of the vector. Due to the sorting, I would never coalesce styles of all the `MarkedTexts` that had identical ranges since they were not adjacent to each other in vector. So, my initial idea was to remove overlapping `MarkedTexts` that had identical ranges as highlights. However, just after testing, this function doesn't do anything anymore because I improved my "coalescing adjacent identical ranges" function and took out the sorting. Sorry for the confusion, I'll remove this piece of code for the next push.

Reply:
If what I said doesn't make any sense, pls let me know and I can explain better with visuals because the way highlights works with the existing code rn is definitely confusing 😞